### DISCUSSION TITLE
How to have the INSTALL working with multiple deployments?

### DISCUSSION THREAD

**Discussion by SMorettini** [nasa/fprime#1710] [Q&A]
Hello, 
I created a structure similar to the one proposed in https://github.com/nasa/fprime/discussions/1638#discussioncomment-3459090 . I aim to have one build folder with multiple deployments and one executable for each deployment.

### My Project structure and content
```
my_proj/
├─ mission/
│  ├─ Simulator/
│  │  ├─ Topology/
│  │  └─ CMakeList.txt(1)
│  ├─ Flight/
│  │  ├─ Topology/
│  │  └─ CMakeLists.txt(2)
│  └─ CMakeList.txt(3)
├─ Fprime/
└─ MyComponents/
```
Content of `CMakeList.txt(3)`:
```
...
# Here I include Fprime and MyComponents
...
add_subdirectory("${CMAKE_CURRENT_LIST_DIR}/Simulator")
add_subdirectory("${CMAKE_CURRENT_LIST_DIR}/Flight")
```
Content of `CMakeList.txt(1)`:
```
add_fprime_subdirectory("${CMAKE_CURRENT_LIST_DIR}/Topology/")
set(SOURCE_FILES "${CMAKE_CURRENT_LIST_DIR}/Topology/Main.cpp")
set(MOD_DEPS Flight/Topology)
register_fprime_deployment()
```
The content of `CMakeList.txt(2)` is almost the same of `CMakeList.txt(1)` but the path of `MOD_DEPS` is changed with the right path.

### The problem

The build works correctly until the INSTALL phase is reached:
```
-- Installing: /home/user/my_proj/mission/build-artifacts/Linux/lib/static/libPro1.a
-- Installing: /home/user/my_proj/mission/build-artifacts/Linux/lib/static/libPro2.a
-- Installing: /home/user/my_proj/mission/build-artifacts/Linux/lib/static/libFlight_Topology.a
-- Installing: /home/user/my_proj/mission/build-artifacts/Linux/dict/FlightTopologyAppDictionary.xml
CMake Error at Simulator/cmake_install.cmake:47 (file):
  file INSTALL cannot find
  "/home/user/my_proj/mission/build-fprime-automatic-native/bin/Linux/Simulator":
  No such file or directory.
Call Stack (most recent call first):
  cmake_install.cmake:52 (include)

make[3]: *** [Makefile:74: install] Error 1
make[2]: *** [Flight/CMakeFiles/Flight.dir/build.make:165: bin/Linux/Flight] Error 2
make[2]: *** Deleting file 'bin/Linux/Flight'
make[1]: *** [CMakeFiles/Makefile2:10642: Flight/CMakeFiles/Flight.dir/all] Error 2
make: *** [Makefile:130: all] Error 2
[ERROR] CMake erred with return code 2
```

### My analysis results and opinions

The error is coming from the `install.cmake` where the `install_add_deployment_target` is not designed to support multiple deployments. The last line of `install_add_deployment_target` is `add_custom_command(TARGET "${MODULE}" POST_BUILD COMMAND "${CMAKE_COMMAND}" --build "${CMAKE_BINARY_DIR}" --target install)`. This is executed as soon as Flight is built but is not waiting for Simulator to be built. 

### My temporary solutions
I commented the `add_custom_command(...)` from `install_add_deployment_target` and I add the following at `CMakeList.txt(3)`
```
...
        add_custom_target( targetAllBuild ALL COMMAND echo "This is a dummy target to check Flight and Simulator are built" DEPENDS Flight Simulator)
        add_custom_command(TARGET targetAllBuild POST_BUILD COMMAND "${CMAKE_COMMAND}" --build "${CMAKE_BINARY_DIR}" --target install)
```
### Questions

1. Is there a better solutions?
2. Am I doing something wrong or Fprime does not support what I want to do?

**Comment by LeStarch** [comment]
Typically we expect deployments to be built individually. That is, you'd build Flight and Simulator separately. Combining them together is not something we've thought through, although it seems like it is something you can *almost* do.

Your solution seems reasonable given the current state of F´.  With a free moment, I will think more deeply about this problem.

**Reply by SMorettini** [reply]
Thanks for the answer. 

I think it could be a nice feature to support. The main motivation for why we went for that was to have an easy development workflow and mainly a simple and faster CI since by building once we have all the executables.
